File: src/pgo/model/tla/TLAConstantDeclaration.java
package pgo.model.tla;

import java.util.List;
import java.util.stream.Collectors;

import pgo.util.SourceLocation;

public class TLAConstantDeclaration extends TLAUnit {
	
	private List<TLAOpDecl> constants;

	public TLAConstantDeclaration(SourceLocation location, List<TLAOpDecl> constants) {
		super(location);
		this.constants = constants;
	}
	
	@Override
	public TLAConstantDeclaration copy() {
		return new TLAConstantDeclaration(getLocation(), constants.stream().map(TLAOpDecl::copy).collect(Collectors.toList()));
	}

	public List<TLAOpDecl> getConstants() {
		return constants;
	}

	@Override
	public <T, E extends Throwable> T accept(TLAUnitVisitor<T, E> v) throws E {
		return v.visit(this);
	}

	@Override
	public int hashCode() {
		final int prime = 31;
		int result = 1;
		result = prime * result + ((constants == null) ? 0 : constants.hashCode());
		return result;
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj)
			return true;
		if (obj == null)
			return false;
		if (getClass() != obj.getClass())
			return false;
		TLAConstantDeclaration other = (TLAConstantDeclaration) obj;
		if (constants == null) {
			return other.constants == null;
		} else return constants.equals(other.constants);
	}

}

File: src/pgo/model/tla/TLAConstantDeclaration.scala
package pgo.model.tla

import pgo.util.SourceLocation

import scala.jdk.CollectionConverters._

final class TLAConstantDeclaration(loc: SourceLocation, val constants: List[TLAOpDecl]) extends TLAUnit(loc) with TLADefinitionComposite {
  def this(loc: SourceLocation, constants: java.util.List[TLAOpDecl]) =
    this(loc, constants.asScala.toList)

  override val definitions: List[TLADefinition] = List(this)

  @Deprecated
  def getConstants: java.util.List[TLAOpDecl] = constants.asJava

  override def copy: TLAConstantDeclaration = new TLAConstantDeclaration(loc, constants.map(_.copy))

  override def accept[T, E <: Throwable](v: TLAUnitVisitor[T, E]): T = v.visit(this)

  override def members: List[TLADefinition] = constants

  override def equals(other: Any): Boolean = other match {
    case that: TLAConstantDeclaration =>
      constants == that.constants
    case _ => false
  }

  override def hashCode(): Int = constants.hashCode
}
